File: 984-most-stones-removed-with-same-row-or-column/most-stones-removed-with-same-row-or-column.py
from collections import defaultdict
class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        discarded = set()
        x_adj = defaultdict(list)
        y_adj = defaultdict(list)

        for i,(x, y) in enumerate(stones):
            x_adj[x].append(i)
            y_adj[y].append(i)
        
        def dfs(i):
            if i not in discarded:
                discarded.add(i)

                for neighbour in x_adj[stones[i][0]]:
                    dfs(neighbour)
                for neighbour in y_adj[stones[i][1]]:
                    dfs(neighbour)

        cnt = 0
        for i in range(len(stones)):
            if i not in discarded:
                dfs(i)
                cnt += 1

        return len(stones) - cnt

File: 984-most-stones-removed-with-same-row-or-column/most-stones-removed-with-same-row-or-column.cpp
class Solution {
public:
    int removeStones(vector<vector<int>>& stones) {
        unordered_set<int> discarded;
        unordered_map<int, vector<int>> x_adj;
        unordered_map<int, vector<int>> y_adj;
        for (int i=0; i<stones.size(); i++) {
            x_adj[stones[i][0]].push_back(i);
            y_adj[stones[i][1]].push_back(i);
        }
        // for (auto i: y_adj) {
        //     cout<<i.first<<": ";
        //     for (auto j: i.second) {
        //         cout<<j<<" ";
        //     }
        //     cout<<endl;
        // }
        
        function<void(int)> dfs = [&](int i) -> void {
            if (!discarded.count(i)) {

                discarded.insert(i);
                for (int j=0; j<x_adj[stones[i][0]].size(); j++) {
                    int neighbour = x_adj[stones[i][0]][j];
                    dfs(neighbour);
                }
                for (int j=0; j<y_adj[stones[i][1]].size(); j++) {
                    int neighbour = y_adj[stones[i][1]][j];
                    dfs(neighbour);
                }
            }
        };
        int cnt = 0;
        for (int i=0; i<stones.size(); i++) {
            if (!discarded.count(i)) {
                dfs(i);
                cnt++;
            }
        }
        return stones.size() - cnt;
    }
};